### PR TITLE
feat: retrieval export

### PR DESCRIPTION
**Pull request by jcace**
Port over function to do exporting of retrieved files, add a cli flag to handle that

### CODE REVIEW

**Comment by elijaharita** [comment]
this is beautiful @jcace love your code! can we just get a test for the export function? doesn't need to be perfect, just at least test that the expected file is created and it's not 0 bytes. 

another thing for consideration (which we can do later if not now) - is it possible to automatically figure out what to name the root output file from ipld metadata?

File: export_test.go [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+	"github.com/urfave/cli/v2"
+)
+
+func TestExportFile(t *testing.T) {
+	outputFilename := "TestExportFile.test.output"
+	// Remove any pre-existing file and ensure it is removed when the test finishes
+	os.Remove(outputFilename)
+	defer func() {
+		os.Remove(outputFilename)
+	}()
```
Comment (jcace) [inline export_test.go]: @elijaharita This seems kind of awkward, is it okay? Do you know of some other builtin way to test functions that output to files, and automatically clean up the files in Go?

File: .gitignore [hunk 2]
```diff
@@ -1,3 +1,4 @@
 build
 extern
-__debug_bin
\ No newline at end of file
+__debug_bin
+*.test.output.*
```
Comment (jcace) [inline .gitignore]: _shouldn't_ be necessary if we clean up after every test, but just wanted to ensure commit history remains clean without any accidental test files in it.

**Comment by jcace** [comment]
> another thing for consideration (which we can do later if not now) - is it possible to automatically figure out what to name the root output file from ipld metadata?

Yup, let me look at that. Might need to rework flags on the `filctl` side as an unspecified filename will simply leave the file in the blockstore and not export it at all

File: export_test.go (hunk 1 above)
Comment (elijaharita) [reply export_test.go]: you can use `t.TempDir()` to get a new empty test directory generated by go which is automatically cleaned up after test finishes running

Comment (elijaharita) [reply export_test.go]: so `outputFilename` would become `path.Join(t.TempDir(), "TestExportFile.test.output")`

what's the purpose of the output file naming scheme you used? is there any reason to do that over something simpler like just `export-test`?

File: .gitignore (hunk 2 above)
Comment (elijaharita) [reply .gitignore]: ah i see why you had test.output extension now. there's no reason this needs to be here though if you use `t.TempDir()`

**Comment by elijaharita** [comment]
> > another thing for consideration (which we can do later if not now) - is it possible to automatically figure out what to name the root output file from ipld metadata?
> 
> Yup, let me look at that. Might need to rework flags on the `filctl` side as an unspecified filename will simply leave the file in the blockstore and not export it at all

ok, that's a good solution for now. 

on this topic, we should later add an export command to just read existing data out of the local blockstore to a file. this could be under a `filctl blockstore` subcommand and have stuff like `filctl blockstore export <cid>`, `filctl blockstore delete <cid>`, `filctl blockstore clear` (instead of `filctl clear-blockstore`), etc.

**Comment by jcace** [comment]
@elijaharita  updated!

Now, `filctl retrieve` will always export to a file, even if no path/name is provided. 

It will simply default to using the CID as the filename, adding .car extension if necessary.

I think this is a good separation of concerns, as the `filctl` command is more user-facing, we can make the assumption that the user will want the resulting exported file. Whereas, code calling the `Retrieve` function may not actually want this so it's fine that they're separated

**Review by elijaharita — APPROVED** [review]

**Pull request merged**